cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAutoRebalancingReconciler.java

```diff
@@ -0,0 +1,530 @@
+
+    private static final ScalingNodes EMPTY_SCALING_NODES = new ScalingNodes(Set.of(), Set.of());
+
+    private final Reconciliation reconciliation;
+    private final Kafka kafkaCr;
+    private final CrdOperator<KubernetesClient, KafkaRebalance, KafkaRebalanceList> kafkaRebalanceOperator;
+    private final Set<Integer> scalingDownBlockedNodes;
+    private final Set<Integer> scalingUpAddedNodes;
+
+    /**
+     * Constructs the Kafka auto-rebalancing reconciler
+     *
+     * @param reconciliation    Reconciliation marker
+     * @param kafkaCr   The Kafka custom resource
+     * @param supplier  Supplies the operators for different resources
+     * @param scalingDownBlockedNodes  nodes blocked on scaling down because of the need for auto-rebalancing first
+     */
+    public KafkaAutoRebalancingReconciler(
+            Reconciliation reconciliation,
+            Kafka kafkaCr,
+            ResourceOperatorSupplier supplier,
+            Set<Integer> scalingDownBlockedNodes) {
+        this.reconciliation = reconciliation;
+        this.kafkaCr = kafkaCr;
+        this.kafkaRebalanceOperator = supplier.kafkaRebalanceOperator;
+        this.scalingDownBlockedNodes = scalingDownBlockedNodes;
+        this.scalingUpAddedNodes = new LinkedHashSet<>();
+    }
+
+    /**
+     * The main reconciliation method which triggers the whole reconciliation pipeline. This is the method which is
+     * expected to be called from the outside to trigger the reconciliation.
+     *
+     * @param kafkaStatus The Kafka Status class for updating auto-rebalancing status on it during the reconciliation
+     *
+     * @return  Future which completes when the reconciliation completes
+     */
+    public Future<Void> reconcile(KafkaStatus kafkaStatus) {
+        // Cruise Control is not defined in the Kafka custom resource or no autorebalance enabled, so nothing to reconcile
+        if (kafkaCr.getSpec().getCruiseControl() == null || kafkaCr.getSpec().getCruiseControl().getAutoRebalance() == null) {
+            LOGGER.infoCr(reconciliation, "Cruise Control or inner autorebalance field not defined in the Kafka custom resource, no auto-rebalancing to reconcile");
+            return Future.succeededFuture();
+        }
+
+        KafkaAutoRebalanceStatus kafkaAutoRebalanceStatus = kafkaCr.getStatus() != null ? kafkaCr.getStatus().getAutoRebalance() : null;
+
+        if (kafkaAutoRebalanceStatus != null) {
+            LOGGER.infoCr(reconciliation, "Loaded auto-rebalance state from the Kafka CR [{}]", kafkaAutoRebalanceStatus.getState());
+            ScalingNodes scalingNodes = getScalingNodes(kafkaStatus.getAutoRebalance());
+            return computeNextStatus(kafkaAutoRebalanceStatus, scalingNodes)
+                    .onComplete(v -> kafkaStatus.setAutoRebalance(kafkaAutoRebalanceStatus));
+        } else {
+            LOGGER.infoCr(reconciliation, "No auto-rebalance state from the Kafka CR, initializing to [Idle]");
```

Comment:
Absolutely I am just taking logging level at INFO for debugging purpose even because with coming feedback I test stuff again and I need them (to avoid going back and forth between DEBUG and INFO each time). I will do a logging level review at the end when all feedback are addressed if it's fine with you.